Port some basic Temporal.PlainDateTime tests from Demitasse to test262

jessealama opened the pull request:
This is the modest beginnings of a larger porting project, moving existing Demitasse for Temporal's `PlainDateTime` to test262.

Much remains to be ported. But in the interest of preserving the sanity of everyone involved, I suggest doing the porting in human-consumable chunks rather than making a single huge PR.

ptomato commented on test/built-ins/Temporal/PlainDateTime/constructor-full.js:
```diff
@@ -0,0 +1,37 @@
+
+TemporalHelpers.assertPlainDateTime(datetime,
+  1976, 11, 'M11', 18, 15, 23, 30, 123, 456, 789,
+  "check instance (all arguments supplied)",
+  undefined,
+  undefined
```
```suggestion
  "check instance (all arguments supplied)"
```

ptomato commented on test/built-ins/Temporal/PlainDateTime/constructor-full.js:
```diff
@@ -0,0 +1,37 @@
+  'calendar supplied in constructor can be extracted and is unchanged'
+);
+
+assert.compareArray(
```
Normally I'm a fan of conciseness but I'd recommend 5x `assert.sameValue` here because otherwise the assertion message will not help pinpoint what the failure is, if there is one.

jessealama replied:
done

jessealama replied:
Would it make sense to extend `assertPlainDateTime` with additional (optional) arguments to check this data? Then these tests could be put into a single function call.

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/hour-undefined.js:
```diff
@@ -5,12 +5,19 @@
 const implicit = new Temporal.PlainDateTime(...args);
-assert.sameValue(implicit.hour, 0, "hour default argument");
+
+TemporalHelpers.assertPlainDateTimesEqual(explicit, implicit, "hour default argument (argument missing)");
```
I'd prefer just using `TemporalHelpers.assertPlainDateTime` here as well; I don't think comparing several objects under test is as useful as testing the expectations directly.

jessealama replied:
done

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/constructor-full.js:
```diff
@@ -0,0 +1,35 @@
+  "calendar supplied in constructor can be extracted and is unchanged"
+);
+
+assert.sameValue(datetime.dayOfWeek, 4, "check day of week information");
+assert.sameValue(datetime.dayOfYear, 323, "check day of year information");
+assert.sameValue(datetime.weekOfYear, 47, "check week of year information");
+assert.sameValue(datetime.daysInWeek, 7, "check number of days in week");
+assert.sameValue(datetime.monthsInYear, 12, "check number of months in year");
+
+assert.sameValue(
+  `${datetime}`,
+  "1976-11-18T15:23:30.123456789",
+  "check string value"
+);
```
I would probably move those to `prototype/dayOfWeek/` etc.

jessealama replied:
done

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/constructor-full.js:
```diff
@@ -0,0 +1,35 @@
+assert.sameValue(datetime.monthsInYear, 12, "check number of months in year");
+
+assert.sameValue(
+  `${datetime}`,
```
`datetime.toString()` is simpler and (imo) more readable.

jessealama replied:
done

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/nanosecond-undefined.js:
```diff
@@ -5,12 +5,19 @@
 const explicit = new Temporal.PlainDateTime(...args, undefined);
-assert.sameValue(explicit.nanosecond, 0, "nanosecond default argument");
+
+TemporalHelpers.assertPlainTimesEqual(
```
I don't think this is right

jessealama replied:
oops, right (copy & paste error)

fixed

Ms2ger replied:
Doesn't seem to be fixed yet

jessealama replied:
oops, right -- done now

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/minute-undefined.js:
```diff
@@ -5,12 +5,19 @@
 const implicit = new Temporal.PlainDateTime(...args);
-assert.sameValue(implicit.minute, 0, "minute default argument");
+
+TemporalHelpers.assertPlainTimesEqual(explicit, implicit, "minute default argument (argument missing)");
```
In any case, this should compare PlainDateTimes, not PlainTimes

jessealama replied:
right, that's clearly a blooper -- I've adopted your suggestion to use `assertPlainDateTime` rather than `assertPlainDateTimesEqual`, thereby fixing this particular issue.

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/prototype/with/cases.js:
```diff
@@ -0,0 +1,77 @@
+
+const datetime = new Temporal.PlainDateTime(1976, 11, 18, 15, 23, 30, 123, 456, 789);
+
+assert.sameValue(
+  `${datetime.with({ year: 2019 })}`,
+  "2019-11-18T15:23:30.123456789",
+  "with year works"
+);
```
I would also prefer `TemporalHelpers.assertPlainDateTime` over comparing stringifications here (and in general, when not specifically testing stringification).

jessealama replied:
done

Ms2ger replied:
Doesn't seem to have made it

jessealama replied:
Looks like I accidentally included `temporalHelpers.js`. Dropped now.

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/prototype/with/options-empty.js:
```diff
@@ -0,0 +1,23 @@
+assert.sameValue(
+  `${datetime.with({ day: 5 }, () => {})}`,
+  expected,
+  "options may be a function that returns an empty object"
```
Note that "that returns an empty object" is not really relevant (and may be confusing). We're just trying to fetch properties off the function object, not calling the function.

jessealama replied:
I've updated the test description -- how does it look now? (Thanks for clarifying -- I didn't realize we weren't calling the function.)

Ms2ger reviewed (commented):
Hope this helps

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/nanosecond-undefined.js:
```diff
@@ -5,12 +5,19 @@
 
-const implicit = new Temporal.PlainDateTime(...args);
-assert.sameValue(implicit.nanosecond, 0, "nanosecond default argument");
+TemporalHelpers.assertPlainTimesEqual(
```
Same

jessealama replied:
right, sorry about that -- fixed now

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/prototype/daysInWeek/basic.js:
```diff
@@ -0,0 +1,12 @@
+
+const calendar = Temporal.Calendar.from("iso8601");
+const datetime = new Temporal.PlainDateTime(1976, 11, 18, 15, 23, 30, 123, 456, 789, calendar);
+assert.sameValue(datetime.dayOfWeek, 4, "check day of week information");
```
```suggestion
assert.sameValue(datetime.daysInWeek, 7, "check days in week information");
```

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/prototype/daysInYear/basic.js:
```diff
@@ -0,0 +1,12 @@
+
+const calendar = Temporal.Calendar.from("iso8601");
+const datetime = new Temporal.PlainDateTime(1976, 11, 18, 15, 23, 30, 123, 456, 789, calendar);
+assert.sameValue(datetime.daysInYear, 366, "check days of year information");
```
```suggestion
assert.sameValue(datetime.daysInYear, 366, "check days in year information");
```

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/prototype/weekOfYear/basic.js:
```diff
@@ -0,0 +1,12 @@
+
+const calendar = Temporal.Calendar.from("iso8601");
+const datetime = new Temporal.PlainDateTime(1976, 11, 18, 15, 23, 30, 123, 456, 789, calendar);
+assert.sameValue(datetime.weekOfYear, 47, "check months of year information");
```
```suggestion
assert.sameValue(datetime.weekOfYear, 47, "check week of year information");
```

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/prototype/with/multiple-unrecognized-properties-ignored.js:
```diff
@@ -0,0 +1,22 @@
+  let plural = `${unit}s`;
+  let arg = { month: 12 };
+  arg[plural] = 1;
+  assert.sameValue(
+    `${datetime.with(arg)}`,
+    "1976-12-18T15:23:30.123456789",
+    `unrecognized property (${plural}) gets ignored`
+  );
```
Please use `TemporalHelpers.assertPlainDateTime` here as well

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainTime/argument-object.js:
```diff
@@ -0,0 +1,35 @@
+  "empty object not an acceptable argument"
+);
+
+assert.sameValue(
+  `${dt.withPlainTime({ hour: 10 })}`,
+  "2015-12-07T10:00:00",
+  "plain object (hour) works"
+);
+
+assert.throws(
+  TypeError,
+  () => dt.withPlainTime({ hours: 9 }), // should be "hour", see above
+  "plain object with a single unrecognized property fails"
+);
+
+assert.sameValue(
+  `${dt.withPlainTime({ hour: 10, seconds: 123 })}`,
+  "2015-12-07T10:00:00",
+  "unrecognized properties are ignored if at least one recognized property is present"
+);
+
```
TemporalHelpers.assertPlainDateTime

jessealama replied:
done

jessealama commented on test/built-ins/Temporal/PlainDateTime/prototype/with/options-empty.js:
```diff
@@ -0,0 +1,23 @@
+  `${datetime.with({ day: 5 }, () => {})}`,
+  expected,
+  "read empty options from function object"
+);
```
done

ptomato commented on test/built-ins/Temporal/PlainDateTime/comparison.js:
```diff
@@ -0,0 +1,64 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
Another naming nitpick, sorry ... "comparison" sounds like it should be testing the compare() method. I'd actually suggest moving this one into the `built-ins/Temporal/PlainDateTime/prototype/valueOf` directory because that is really what it's testing. Check `built-ins/Temporal/PlainDate/prototype/valueOf/basic.js` for reference (although your way is more explicit with the assertion messages, which is good.)

jessealama replied:
sure, sounds good -- done

ptomato commented on test/built-ins/Temporal/PlainDateTime/constructor-full.js:
```diff
@@ -0,0 +1,29 @@
+  "calendar supplied in constructor can be extracted and is unchanged"
+);
+
+assert.sameValue(
```
Consider separating this test into `prototype/toString/basic.js`.

jessealama replied:
done

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/add/hour-overflow.js:
```diff
@@ -0,0 +1,25 @@
+TemporalHelpers.assertPlainDateTime(
+  later.add({ hours: -12 }),
+  2019, 10, "M10", 28, 22, 46, 38, 271, 986, 102,
+  "symmetrical with regard to negative durations"
```
This assertion message seems wrong.

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/equals/argument-object-may-throw.js:
```diff
@@ -0,0 +1,16 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
Same comment as above about the name

jessealama replied:
done

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/equals/normal-case.js:
```diff
@@ -0,0 +1,14 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
Same comment as above about the name (I suggest `basic.js`)

jessealama replied:
done

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/subtract/hour-overflow.js:
```diff
@@ -0,0 +1,33 @@
+TemporalHelpers.assertPlainDateTime(
+  earlier.add({ hours: -12 }),
+  2019, 10, "M10", 28, 22, 46, 38, 271, 986, 102,
+  "symmetrical with regard to negative durations (1)"
```
Same comment as above about the assertion message

jessealama replied:
fixed

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/subtract/hour-overflow.js:
```diff
@@ -0,0 +1,33 @@
+TemporalHelpers.assertPlainDateTime(
+  later2.subtract({ hours: -2 }),
+  2020, 6, "M06", 1, 1, 12, 38, 271, 986, 102,
+  "symmetrical with regard to negative durations (2)"
```
Ditto

jessealama replied:
done

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/with/cases.js:
```diff
@@ -0,0 +1,77 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
Same comment as above about the name, I'd recommend `basic.js`

jessealama replied:
done

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/with/misspelled-property.js:
```diff
@@ -0,0 +1,23 @@
+  assert.throws(
+    TypeError,
+    () => instance.with(options),
+    `plural unit ("${plural}" vs "${unit}") is not acceptable`
```
You could consider combining this with the previous empty-object test under the banner "at least one correctly-spelled property must be present". (and then also test a nonsense property such as `"other string"`)

jessealama replied:
done

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainDate/argument-string-non-iso-calendar.js:
```diff
@@ -0,0 +1,23 @@
+assert.sameValue(
+  shifted.calendar.toString(),
+  "japanese",
+  "result contains a non-ISO calendar if present in the input (1)"
```
```suggestion
  "result contains a non-ISO calendar if present in the input (2)"
```
?

jessealama replied:
done

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainDate/unrecognized-properties.js:
```diff
@@ -0,0 +1,23 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
It's more convenient for keeping track of what coverage there is, if you can pattern similar tests in different methods similarly, and also name them similarly.

For example, the tests for with() have the empty object, objects with only unrecognized properties, and objects with both recognized and unrecognized properties, as three different tests (and I suggested combining the former two), whereas this file combines the latter two, and withPlainTime() combines all three.

jessealama replied:
makes sense -- I've tried to clean things up as suggested

ptomato reviewed (commented):
Thanks! Some more comments, most of them about naming...

jessealama commented on test/built-ins/Temporal/PlainDateTime/compare/object-argument-throws.js:
```diff
@@ -0,0 +1,23 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
done

jessealama commented on test/built-ins/Temporal/PlainDateTime/compare/normal-case.js:
```diff
@@ -0,0 +1,29 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
ah, OK -- done

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/subtract/casts-argument.js:
```diff
@@ -0,0 +1,23 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
Same comments for this file as for the one under `add/`.

jessealama replied:
done

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/subtract/overflow-mixed-throws.js:
```diff
@@ -0,0 +1,18 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
Same comment here as in the corresponding test under `add/`.

jessealama replied:
done I think?

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/until/argument-cast.js:
```diff
@@ -0,0 +1,20 @@
+
+const datetime = new Temporal.PlainDateTime(1976, 11, 18, 15, 23, 30, 123, 456, 789);
+
+const duration1 = datetime.until({ year: 2019, month: 10, day: 29, hour: 10 });
```
These aren't used below.

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/until/argument-cast.js:
```diff
@@ -0,0 +1,20 @@
+const duration1 = datetime.until({ year: 2019, month: 10, day: 29, hour: 10 });
+const duration2 = datetime.until("2019-10-29T10:46:38.271986102");
+
+TemporalHelpers.assertDurationsEqual(
```
Is this test going to pass? The property bag argument and the string argument don't denote the same date-time, so it doesn't seem like the result should be the same.

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/until/different-calendars-throws.js:
```diff
@@ -0,0 +1,17 @@
+---*/
+
+const dt1 = new Temporal.PlainDateTime(2000, 1, 1, 0, 0, 0, 0, 0, 0);
+const dt2 = new Temporal.PlainDateTime(2000, 1, 1, 0, 0, 0, 0, 0, 0, Temporal.Calendar.from("japanese"));
```
This line will fail on non-ECMA-402 hosts, so it should use a custom calendar instead.

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/until/returns-days.js:
```diff
@@ -0,0 +1,44 @@
+const feb20 = new Temporal.PlainDateTime(2020, 2, 1, 0, 0);
+const feb21 = new Temporal.PlainDateTime(2021, 2, 1, 0, 0);
+
+TemporalHelpers.assertDuration(
```
There's some overlap between this file and the previous one.

jessealama replied:
Hmm, the previous test doesn't explicitly test `largestUnit` = `auto`, so to my eyes this seems at least superficially different. Perhaps I'm missing something?

Ms2ger replied:
I'm guessing "some overlap" doesn't mean "completely superfluous" - I think it's fine to leave this as is, and I'll take another look at these tests next week.

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/until/subseconds.js:
```diff
@@ -0,0 +1,36 @@
+
+/*---
+esid: sec-temporal.plaindatetime.prototype.until
+description: String argument gets cast
```
This description doesn't seem correct for the test.

jessealama replied:
Looks like a copy & paste relic. I've tried to improve it now.

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/until/units-changed.js:
```diff
@@ -0,0 +1,42 @@
+
+/*---
+esid: sec-temporal.plaindatetime.prototype.until
+description: Units may be changed
```
I'm not sure I understand this description, what does this test show? Is it a full test of the `largestUnit` option?

jessealama replied:
I think it's just a bunch of basic cases, showing off that `largestUnit` is respected. I can add some more cases to make it a test of the complete range of acceptable units.

Ms2ger replied:
```suggestion
description: Largest unit is respected
```

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/until/weeks-months-mutually-exclusive.js:
```diff
@@ -0,0 +1,24 @@
+
+TemporalHelpers.assertDuration(
+  dt.until(laterDateTime, { largestUnit: "weeks" }),
+  0, 0, 1, 0, 0, 0, 0, 0, 0,
```
I'm not sure these values are correct, 42 days can't be 1 week.

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/until/weeks-months-mutually-exclusive.js:
```diff
@@ -0,0 +1,24 @@
+
+TemporalHelpers.assertDuration(
+  dt.until(laterDateTime, { largestUnit: "months" }),
+  0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
```
Likewise, this seems like it should be 1 month and several days and hours.

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainDate/argument-string-iso-calendar.js:
```diff
@@ -0,0 +1,23 @@
+---*/
+
+const dt = new Temporal.PlainDateTime(1995, 12, 7, 3, 24, 30);
+const shifted = dt.withPlainDate("2008-09-06[u-ca=japanese]");
```
This will fail on non-ECMA-402 hosts. But it's not possible to create a PlainDate from a string with a custom calendar, so maybe this test should be moved to `intl402/`.

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainDate/argument-string-non-iso-calendar.js:
```diff
@@ -0,0 +1,23 @@
+---*/
+
+const dt = new Temporal.PlainDateTime(1995, 12, 7, 3, 24, 30);
+const shifted = dt.withCalendar("japanese").withPlainDate("2008-09-06");
```
This will fail on non-ECMA-402 hosts, so it should use a custom calendar.

jessealama replied:
done

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainDate/non-compatible-calendars-throw.js:
```diff
@@ -0,0 +1,16 @@
+
+assert.throws(
+  RangeError,
+  () => dt.withCalendar("gregory").withPlainDate("2008-09-06[u-ca=japanese]"),
```
This will fail on non-ECMA-402 hosts, so it should use a custom calendar.

jessealama replied:
Could we keep it as-is, but move it to the `intl/` directory?

ptomato replied:
If we did that, the functionality (non-compatible calendars in withPlainDate must throw) wouldn't get tested at all on non-ECMA-402 hosts.

jessealama replied:
fixed by using a fake calendar

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainTime/argument-object.js:
```diff
@@ -0,0 +1,35 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
This test should be named the same as the other similar tests for other methods.

jessealama replied:
done

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainTime/no-argument-default-to-midnight.js:
```diff
@@ -0,0 +1,26 @@
+
+/*---
+esid: sec-temporal.plaindatetime.prototype.withplaintime
+description: IF no argument is given, default to midnight
```
```suggestion
description: If no argument is given, default to midnight
```

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainTime/no-argument-default-to-midnight.js:
```diff
@@ -0,0 +1,26 @@
+  12,
+  "M12",
+  7,
+    0,
```
```suggestion
  0,
```

jessealama commented:
> Thanks. At this point it might be efficient to add further tests to a new PR after addressing the comments on this one. It's fairly time consuming to view all the files to see which ones have changed.

I agree. I'll focus on this PR and continue the porting task in a separate PR.

jessealama commented on test/built-ins/Temporal/PlainDateTime/prototype/equals/cast.js:
```diff
@@ -0,0 +1,38 @@
+const dt2 = new Temporal.PlainDateTime(2019, 10, 29, 10, 46, 38, 271, 986, 102);
+
+assert.sameValue(
+  dt1.equals({ year: 1976, month: 11, day: 18, hour: 15 }),
```
I need to fix a bunch of failing tests...sorry! This is just one of several failing tests.

jessealama replied:
(Yet somehow the CI pipelines were all green?)

ptomato replied:
The CI pipelines only report whether each host's runner was able to execute the tests. They don't actually report whether the tests failed or succeeded on each host. This is to accommodate proposals, where we add tests in advance of hosts implementing them, they're expected to fail. It's our responsibility to make sure the tests pass the Temporal polyfill :smile:

jessealama replied:
got it!

rwaldron commented on test/built-ins/Temporal/PlainDateTime/compare/argument-object-throws-insufficient-data.js:
```diff
@@ -0,0 +1,23 @@
+
+assert.throws(
+  TypeError,
+  () => PlainDateTime.compare({ year: 1976 }, dt2),
```
Should this be `Temporal.PlainDateTime`?

jessealama replied:
yes, thanks! (This, along with several other bloopers, need to be fixed on my end.)

jessealama commented on test/built-ins/Temporal/PlainDateTime/prototype/add/casts-argument.js:
```diff
@@ -0,0 +1,23 @@
+  "Duration argument"
+);
+
+TemporalHelpers.assertPlainDateTime(
```
true, good point -- this test can be removed

jessealama commented on test/built-ins/Temporal/PlainDateTime/prototype/add/overflow-wrong-type.js:
```diff
@@ -25,3 +25,12 @@ TemporalHelpers.checkStringOptionWrongType("overflow", "constrain",
   (result, descr) => TemporalHelpers.assertPlainDateTime(result, 2003, 8, "M08", 5, 15, 0, 0, 0, 0, 0, descr),
 );
+
+const badOverflows = ['', 'CONSTRAIN', 'balance', 3, null];
```
done

jessealama commented on test/built-ins/Temporal/PlainDateTime/prototype/add/overflow-mixed-throws.js:
```diff
@@ -0,0 +1,18 @@
+
+/*---
+esid: sec-temporal.plaindatetime.prototype.add
+description: Positive and negative values in overthrow are not acceptable
```
renamed and cleaned up the overthrow

jessealama commented on test/built-ins/Temporal/PlainDateTime/prototype/add/casts-argument.js:
```diff
@@ -0,0 +1,23 @@
+const jan31 = new Temporal.PlainDateTime(2020, 1, 31, 15, 0);
+
+TemporalHelpers.assertPlainDateTime(
+  jan31.add(new Temporal.Duration("P1MT1S")),
```
oops, it wasn't my intent to do something weird. I've replaced the construct with a call to the `from` static method (which was my intent).

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/prototype/with/options-empty.js:
```diff
@@ -0,0 +1,23 @@
+const datetime = new Temporal.PlainDateTime(1976, 11, 18, 15, 23, 30, 123, 456, 789);
+
+TemporalHelpers.assertPlainDateTime(
+  datetime.with({ day: 5 }, {}),
```
More interesting with:
```suggestion
  datetime.with({ day: 40 }, {}),
```

Same below.

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainDate/argument-plaindate.js:
```diff
@@ -0,0 +1,27 @@
+
+TemporalHelpers.assertPlainDateTime(
+  dt.withPlainDate(date),
+  2020,
+  1,
+  "M01",
+  23,
+  3,
+  24,
+  30,
+  0,
+  0,
+  0,
```
Let's move this onto one line

jessealama replied:
done

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainDate/argument-plaindate.js:
```diff
@@ -0,0 +1,27 @@
+
+/*---
+esid: sec-temporal.plaindatetime.withplaindate
+description: Plain object may be acceptable
```
```suggestion
description: PlainDate object is acceptable
```

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainDate/argument-string-iso-calendar.js:
```diff
@@ -0,0 +1,36 @@
+includes: [temporalHelpers.js]
+---*/
+
+const dt = new Temporal.PlainDateTime(1995, 12, 7, 3, 24, 30);
+const cal = {
+  id: 'thisisnotiso',
+  era() { return "the era"; },
+  eraYear() { return 1909; },
+  toString() { return "this is a string"; },
+  year() { return 2008; },
+  month() { return 9; },
+  monthCode() { return "M09"; },
+  day() { return 6; }
+};
+const shifted = dt.withCalendar(cal).withPlainDate("2008-09-06");
```
I found the use of `withCalendar` confusing here - could you pass `cal` to the constructor instead?

jessealama replied:
done

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainDate/argument-string-iso-calendar.js:
```diff
@@ -0,0 +1,36 @@
+  monthCode() { return "M09"; },
+  day() { return 6; }
+};
+const shifted = dt.withCalendar(cal).withPlainDate("2008-09-06");
```
The argument passed here is the same as what your calendar is going to return anyway - let's use another argument.

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainDate/argument-string-iso-calendar.js:
```diff
@@ -0,0 +1,36 @@
+);
+
+assert.sameValue(
+  shifted.calendar.toString(),
+  "this is a string",
```
I think we require the stronger:
```suggestion
  shifted.calendar,
  cal,
```

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainDate/argument-string-non-iso-calendar.js:
```diff
@@ -0,0 +1,36 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
This doesn't appear to be different from `test/built-ins/Temporal/PlainDateTime/prototype/withPlainDate/argument-string-iso-calendar.js`, which makes sense, because you can't get a custom calendar from an ISO string. Let's just remove this.

jessealama replied:
trashed

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainDate/non-compatible-calendars-throw.js:
```diff
@@ -0,0 +1,26 @@
+
+assert.throws(
+  RangeError,
+  () => dt.withCalendar(cal).withPlainDate(new Temporal.PlainDate(2008, 9, 6, anotherCal)),
```
This is a little too brittle: it would pass if `ẁithCalendar` or the constructor would somehow throw a RangeError. Let's define both objects outside this call (and let's pass `cal` to the constructor directly).

jessealama replied:
done

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainTime/argument-string-without-time-designator.js:
```diff
@@ -0,0 +1,26 @@
+
+TemporalHelpers.assertPlainDateTime(
+  dt.withPlainTime("12:34"),
+  2015,
+  12,
+  "M12",
+  7,
+  12,
+  34,
+  0,
+  0,
+  0,
+  0,
```
One line

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainTime/argument-time.js:
```diff
@@ -0,0 +1,29 @@
+
+TemporalHelpers.assertPlainDateTime(
+  dt.withPlainTime(time),
+  2015,
+  12,
+  "M12",
+  7,
+  hour,
+  minute,
+  0,
+  0,
+  0,
+  0,
```
one line

jessealama replied:
done

jessealama commented on test/built-ins/Temporal/PlainDateTime/prototype/add/casts-argument.js:
```diff
@@ -0,0 +1,17 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
done

jessealama commented on test/built-ins/Temporal/PlainDateTime/prototype/add/constrain-ambiguous.js:
```diff
@@ -0,0 +1,29 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
done

jessealama commented on test/built-ins/Temporal/PlainDateTime/prototype/add/options-empty.js:
```diff
@@ -0,0 +1,23 @@
+
+TemporalHelpers.assertPlainDateTime(
+  jan31.add({ years: 1 }, {}),
+  2021, 1, "M01", 31, 15, 0, 0, 0, 0, 0,
```
done

jessealama commented on test/built-ins/Temporal/PlainDateTime/prototype/subtract/constrain-ambiguous.js:
```diff
@@ -0,0 +1,29 @@
+// Copyright (C) 2022 Igalia, S.L. All rights reserved.
```
done

jessealama commented on test/built-ins/Temporal/PlainDateTime/prototype/subtract/options-empty.js:
```diff
@@ -0,0 +1,23 @@
+const jan31 = new Temporal.PlainDateTime(2020, 1, 31, 15, 0);
+
+TemporalHelpers.assertPlainDateTime(
+  jan31.subtract({ years: 1 }, {}),
```
done

Ms2ger commented on test/built-ins/Temporal/PlainDateTime/prototype/subtract/argument-plain-object-mixed-signs.js:
```diff
@@ -0,0 +1,18 @@
+
+/*---
+esid: sec-temporal.plaindatetime.prototype.subtract
+description: Positive and negative values in overthrow are not acceptable
```
```suggestion
description: Positive and negative values in the temporalDurationLike argument are not acceptable
```

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/add/overflow-invalid-string.js:
```diff
@@ -20,4 +20,12 @@ features: [Temporal]
+  assert.throws(
+    RangeError,
+    () => datetime.add({ months: 1 }, { overflow }),
+    `invalid overflow (${typeof overflow})`
```
```suggestion
    `invalid overflow ("${overflow}")`
```
(they're all strings, so otherwise the assertion message wouldn't distinguish them)

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainDate/argument-object-unrecognized-properties.js:
```diff
@@ -0,0 +1,35 @@
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-temporal.plaindatetime.withplaindate
```
```suggestion
esid: sec-temporal.plaindatetime.prototype.withplaindate
```

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainDate/argument-object.js:
```diff
@@ -0,0 +1,17 @@
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-temporal.plaindatetime.withplaindate
```
```suggestion
esid: sec-temporal.plaindatetime.prototype.withplaindate
```

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainDate/argument-plaindate.js:
```diff
@@ -0,0 +1,18 @@
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-temporal.plaindatetime.withplaindate
```
```suggestion
esid: sec-temporal.plaindatetime.prototype.withplaindate
```

ptomato commented on test/built-ins/Temporal/PlainDateTime/prototype/withPlainDate/non-compatible-calendars-throw.js:
```diff
@@ -0,0 +1,29 @@
+};
+
+const date = new Temporal.PlainDate(2008, 9, 6, anotherCal);
+const dtWithNewCalendar = dt.withCalendar(cal);
```
You could dispense with the `withCalendar` call and just create `dt` here with `cal` already in the constructor arguments.

jessealama replied:
done

ptomato reviewed (dismissed):
Looking good. One more naming request, I'd prefer to pick one name for `argument-object-throws-insufficient-data`, `argument-object-insufficient-data`, and `argument-object-unrecognized-properties` and use that consistently.

jessealama commented:
> Looking good. One more naming request, I'd prefer to pick one name for `argument-object-throws-insufficient-data`, `argument-object-insufficient-data`, and `argument-object-unrecognized-properties` and use that consistently.

I've settled on `argument-object-insufficient-data`.

To my eyes, the "unrecognized properties" tests have a different intention than the "insufficient data" tests. I can't deny, though, that the two kinds of tests overlap (and may even exercise the exact same code paths), despite having different origins. In any case, I've done the suggested renaming.
